labs/a10.md

```diff
@@ -0,0 +1,371 @@
+    executable that produces the output you are currently reading.
+ 4. The Docker daemon streamed that output to the Docker client, which sent it
+    to your terminal.
+
+To try something more ambitious, you can run an Ubuntu container with:
+ $ docker run -it ubuntu bash
+
+Share images, automate workflows, and more with a free Docker ID:
+ https://cloud.docker.com/
+
+For more examples and ideas, visit:
+ https://docs.docker.com/engine/userguide/
+ ```
+
+Some [quick definitions from Docker's website:](https://docs.docker.com/get-started/#a-brief-explanation-of-containers)
+
+An **image** is a lightweight, stand-alone, executable package that includes everything needed to run a piece of
+software, including the code, a runtime, libraries, environment variables, and config files. Images are useful primarily for their
+speed, but images can also be used as a base to be built on top of in future images, as you'll see later with Dockerfiles. In the
+last example hello-world was the image used to test our docker installation.
+
+A **container** is a runtime instance of an image—what the image becomes in memory when actually executed.
+It runs completely isolated from the host environment by default, only accessing host files and ports if configured to
+do so. A container gets created upon executing docker run on an image.
+
+This is similar to the distinction between Objects and Classes in Object Oriented Programming. Images would be Classes,
+and containers would be Objects.
+
+Be sure to read through the output from running the hello-world image to get an understanding of what the Docker daemon
+was doing.
+
+### Running an interactive container
+We're now going to walk you through running a container interactively. This is useful if you ever need to play around
+and install stuff on a bare system without messing up your current system. Try running the following command:
+
+`docker run -it ubuntu:xenial /bin/bash`
+
+The `-i` flag tells docker to keep `STDIN` open to your container, and the `-t` flag allocates a
+[pseudo-TTY](https://en.wikipedia.org/wiki/Pseudoterminal) for you. Basically you need both for you to have a way to
+enter text and have this display properly. At the end of the command, `/bin/bash` is just the command you want to run
+once the container starts up. Try installing some packages from `apt` or just play around. It should look like a bare
+Linux system.
+
+You can exit the container with `CTRL+D`.
+
+Notice how even though your VM is running Debian, you were able to run the Xenial version of Ubuntu in a container. If
+you are curious about other variants of Linux, you can run a lot of them inside containers as well! This all works because
+Linuxdistributions all share the Linux kernel. For that same reason, you won't be able to run MacOS or Windows in a container.
+You can try running [Fedora](https://getfedora.org/) (_\*tips hat\*_ M'Linux), another long-running Linux distribution.:
```

Comment:
This is A+ memery, :100:

(nit: also remove the period near the end since it looks strange with a colon following it)